### ISSUE TITLE
Colocation bug for distributed training(NMT model)

### ISSUE THREAD

**Issue by sj6077** [tensorflow/tensorflow#20074]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
Source
- **TensorFlow version (use command below)**:
1.6.1
- **Python version**: 
2.7
- **Bazel version (if compiling from source)**:
0.13.0
- **GCC/Compiler version (if compiling from source)**:
5.4.0
- **CUDA/cuDNN version**: 
9/7
- **GPU model and memory**:
TitanXP/ 12G
- **Exact command to reproduce**:

### Describe the problem
I saw the lines in [NMT](https://github.com/tensorflow/nmt/blob/365e7386e6659526f00fa4ad17eefb13d52e3706/nmt/model_helper.py#L310). If embedding parameter is set in parameter server, embedding_lookup must follow the device placement but the result is opposite(embedding parameter follows embedding_lookup). I think the placement of colocation group in `placer.cc` is wrong so I tried to fix it but it's complicated. Do you have any plan to fix it? or Can you give a suggestion?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
@azaks2 any thoughts?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @azaks2: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by azaks2** [comment]
Any easy way to reproduce the problem? Also what are the downsides for your use case to have > 1 partitions.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @azaks2: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sj6077** [comment]
I'm building an automatic parallelization module on TensorFlow.(https://github.com/snuspl/parallax). However,  users must remember the number of partitions > 1 for NMT model. I want to provide an automatic way that users don't need to care about additional stuff for distributed environment.  Even though I assigned the embedding parameters on a PS, it goes away to worker by the default placer. Is there any way that I can force to assign the parameters to a PS?

**Comment by azaks2** [comment]
model_device_fn = tf.train.replica_device_setter(10)
 with tf.device(model_device_fn):
   ...
   tf.get_variable()

should work. I also tried tf.fixed_size_partitioner(1) and it worked with nmt (tf.get_default_graph() shows expected devices).

**Comment by sj6077** [comment]
Yes, the device is correct but the device placement is changed when running the graph in a session. I checked it log_device_placement=True for session config. I cannot control the final device placement using tf.device() API.

**Comment by azaks2** [comment]
I poked a little more. I could not reproduce the problem. Meaning I always get lookup and the embedding var to be on ps. I did however observed the difference in the placement of dynamic_seq2seq/encoder/rnn/TensorArrayUnstack/TensorArrayScatter/TensorArrayScatterV3 op, which moves from worker to ps whenever we have 1 partition. The reason is that the embedding var is on ps, lookup is collocate with the embedding var, and the lookup result is fed to the tensor array op that wants to be collocated with its input (this can be a performance bug on its own). When num_partitions > 1 there is an intermediate node after the lookup (namely ParallelDynamicStitch) that has no collocation constraints and is placed on the worker. So now tensor array ops can be on the worker and lookup on ps.

Again, I did not witness with tf.device assignments being ignored though internally added collocation constraints can be very intuitive and potentially harmful. To keep things more consistent between having 1 partition or more I would try wrapping embedding lookup with tf.identity. Indeed I am thinking to change embedding_lookup op itself to just always do it on its own.
